### PR TITLE
AAP-62657 Implement logic to extract and populate JWT claims from Controller Jobs

### PR DESCRIPTION
**Pull request by melissalkelly**
AAP-62657

  ## What is being changed?                                                                        
                                                                                                   
  Adds `populate_claims_for_workload()` function to `awx/main/tasks/jobs.py`:                      
                                                                                                   
  - **Extracts 21 JWT claims** - Populates all claims defined in `AutomationControllerJobScope`    
  from `django-ansible-base`                                                                       
  - **Uses DAB constants** - References claim names via `AutomationControllerJobScope.CLAIM_*` for 
  consistency with Gateway                                                                         
  - **Handles null/missing fields** - Omits claims when workload attributes are null or empty      
  strings                                                                                          
  - **Comprehensive tests** - Includes 4 parametrized test cases covering complete jobs, minimal   
  jobs, empty strings, and edge cases                                                              
                                                                                                   
  ## Why is this change needed?                                                                    
                                                                                                   
  This function is the Controller-side implementation for OIDC workload identity claim population. 
  While the scope is defined in django-ansible-base (AAP-62105) and validated by Gateway           
  (AAP-62534), Controller needs to provide the actual job metadata.                                
                                                                                                   
  The function will be called from `RunJob.build_jwt()` (to be implemented in **AAP-62693**) to:   
  - Extract job metadata into the standardized claim format                                        
  - Pass claims to the Workload Identity Client                                                    
  - Enable Gateway to issue signed identity tokens for running jobs                                
                                                                                                   
  ## How does this change address the issue?                                                       
                                                                                                   
  - All 21 claims populated                                                  
  - Null handling implemented                                                                                                                        
  - Tests verify behavior                                                              
                                                                                                   
  ## Notes                                                                                         
                                                                                                   
  This function is **not yet called** - it will be integrated when `RunJob.build_jwt()` is         
  implemented in AAP-62693. The function signature and behavior are complete and tested, ready for 
  integration.                                                                                     
                                                                                                   
  ## Related Tickets                                                                               
                                                                                                   
  - **AAP-62657** (this ticket) - Implement claim population logic                                 
  - **AAP-62693** - Integrate Workload Identity Client to request JWTs on job launch (will call    
  this function)                                                                                   
  - **AAP-62105** - Define `AutomationControllerJobScope` in django-ansible-base                   
  - **AAP-62534** - Gateway validates claims against scope
##### ISSUE TYPE
 - New or Enhanced Feature

##### COMPONENT NAME

 - Other

### CODE REVIEW

**Comment by melissalkelly** [comment]
CI is failing with `ModuleNotFoundError: No module named 'ansible_base.lib.workload_identity'`                                                                              
  - This PR depends on https://github.com/ansible/django-ansible-base/pull/923 (AAP-62105) which creates 'ansible_base.lib.workload_identity'        
  - AAP-62105 was merged to `devel` on **January 29, 2025**                                        
  - CI is using a cached version of django-ansible-base from before the merge

File: awx/main/tasks/jobs.py [hunk 1]
```diff
@@ -92,9 +92,63 @@
 # Django flags
 from flags.state import flag_enabled
 
+# Workload Identity
+from ansible_base.lib.workload_identity.controller import AutomationControllerJobScope
+
 logger = logging.getLogger('awx.main.tasks.jobs')
 
 
+def populate_claims_for_workload(workload):
```
Comment (PabloHiro) [inline awx/main/tasks/jobs.py]:
```suggestion
def populate_claims_for_job(job):
```

Since it is specific to jobs (not e.g. project updates). Thoughts?

File: awx/main/tasks/jobs.py [hunk 2]
```diff
@@ -92,9 +92,63 @@
...
+        """Safely traverse object attributes and dictionary keys."""
+        current = workload
+        for attr in attrs:
+            if current is None:
+                return None
+            if isinstance(current, dict):
+                current = current.get(attr)
+            else:
+                current = getattr(current, attr, None)
+        return current
```
Comment (PabloHiro) [inline awx/main/tasks/jobs.py]: I feel like we can trim this a little bit. Python has a builtin `.get` for dictionaries. Think we can refactor this so we check if `current` is a dict once and then use the builtin method?

**Review by PabloHiro — APPROVED** [review]
LGTM. Please engage with AWX codeowners before merging 🙏

File: awx/main/tests/unit/tasks/test_jobs.py [hunk 3]
```diff
@@ -188,3 +189,106 @@ def test_invalid_host_facts(mock_facts_settings, bulk_update_sorted_by_id, priva
...
+                AutomationControllerJobScope.CLAIM_JOB_NAME: 'Test',
+                AutomationControllerJobScope.CLAIM_JOB_TYPE: 'run',
+                AutomationControllerJobScope.CLAIM_LAUNCH_TYPE: 'manual',
+                AutomationControllerJobScope.CLAIM_ORGANIZATION_ID: 1,
+            },
+        ),
+    ],
+)
+def test_populate_claims_for_job(job_attrs, expected_claims):
+    job = mock.MagicMock(spec=Job)
```
Comment (fosterseth) [inline awx/main/tests/unit/tasks/test_jobs.py]: we have a "job" fixture that is creates a Job django object. Might be more thorough to use that, which would help catch regressions if the job fields change one day

for example, this test uses the fixture https://github.com/fosterseth/awx/blob/292262e14fc66883d97b46f84a9ef891c153e2b6/awx/main/tests/unit/test_tasks.py#L388

**Comment by AlanCoding** [comment]
Is this going to be merged into devel or put into some other feature branch?

It's weird to write this to work with a dict passed in. But that statement is impossible to evaluate in isolation. Maybe you did have a good reason for this to expect a dict instead of a `Job` based on a super opinionated call pattern from an external library that we have no control over. Absent that, I would push to write this to work more directly with the existing models.

**Comment by melissalkelly** [comment]
> Is this going to be merged into devel or put into some other feature branch?

This is planned to be merged into devel. 

> It's weird to write this to work with a dict passed in. But that statement is impossible to evaluate in isolation. Maybe you did have a good reason for this to expect a dict instead of a `Job` based on a super opinionated call pattern from an external library that we have no control over. Absent that, I would push to write this to work more directly with the existing models.

The dict handling is only needed because the job.launched_by property returns a dict like {'id': 10, 'name': 'admin'} rather than a User model object. This is defined in the existing codebase at awx/main/models/unified_jobs.py:1644.

**Comment by AlanCoding** [comment]
Well then, even on the specific point of `launched_by`, I'm going to make to make the point that `CLAIM_LAUNCHED_BY_USER_NAME` misleads. Following the definition of the property, it _might_ be something that is the name of a job or an internally-generated string. Not a username. Perhaps we can split hairs over "username" versus "user name", but my point is that this is not a legitimate field. Sure, I get providing the launched_by details, because that is shown in the API (as we can't take it back, even though we should). I would suggest `CLAIM_LAUNCHED_BY_NAME`, because it's ambiguous, because the data you're providing is ambiguous.

Fine to have an exception for that case, but I'd like to keep the rest of the logic written normally as one would otherwise be expected to interact with a `UnifiedJob` python object.

**Comment by PabloHiro** [comment]
> Well then, even on the specific point of launched_by, I'm going to make to make the point that CLAIM_LAUNCHED_BY_USER_NAME misleads. Following the definition of the property, it might be something that is the name of a job or an internally-generated string. Not a username. Perhaps we can split hairs over "username" versus "user name", but my point is that this is not a legitimate field. Sure, I get providing the launched_by details, because that is shown in the API (as we can't take it back, even though we should). I would suggest CLAIM_LAUNCHED_BY_NAME, because it's ambiguous, because the data you're providing is ambiguous.


@AlanCoding , if I am following, you suggest renaming both `CLAIM_LAUNCHED_BY_USER_NAME` and `CLAIM_LAUNCHED_BY_USER_ID`, with `CLAIM_LAUNCHED_BY_NAME` and `CLAIM_LAUNCHED_BY_ID` correct? Because some jobs can be triggered by non-users right?

@melissalkelly I think this is fair feedback, think you can update it in DAB?
https://github.com/ansible/django-ansible-base/blob/316f7b66c36124ac6af6626551076b31ce09d998/ansible_base/lib/workload_identity/controller.py#L26-L27

>Fine to have an exception for that case, but I'd like to keep the rest of the logic written normally as one would otherwise be expected to interact with a UnifiedJob python object.

We are still figuring out if what we have here is common to all "Job-like" resources in AWX or not. We would need these for every "Job-like" resource in AWX that consumes credentials. Knowing this, do you suggest we place this in a different file? If yes, which?

**Comment by melissalkelly** [comment]
> @melissalkelly I think this is fair feedback, think you can update it in DAB? https://github.com/ansible/django-ansible-base/blob/316f7b66c36124ac6af6626551076b31ce09d998/ansible_base/lib/workload_identity/controller.py#L26-L27
> 
Updated. Please see the PR https://github.com/ansible/django-ansible-base/pull/929
> 
> We are still figuring out if what we have here is common to all "Job-like" resources in AWX or not. We would need these for every "Job-like" resource in AWX that consumes credentials. Knowing this, do you suggest we place this in a different file? If yes, which?

I added unit test coverage for `populate_claims_for_workload` with other workload types that consume credentials (ProjectUpdate, InventoryUpdate, AdHocCommand). To my knowledge these three workloads and Job are the only controller workloads that actually consume credentials. This means that `populate_claims_for_workload` can be used across these workload types. Please share any thoughts or feedback.

**Comment by PabloHiro** [comment]
> I added unit test coverage for `populate_claims_for_workload` with other workload types that consume credentials (ProjectUpdate, InventoryUpdate, AdHocCommand). To my knowledge these three workloads and Job are the only controller workloads that actually consume credentials. This means that `populate_claims_for_workload` can be used across these workload types. Please share any thoughts or feedback.

Thank you for this @melissalkelly , this proves (via tests!) that you were right from the beginning and this was generic enough.

Only question open for me is if the location of this function and tests are on the right file. @AlanCoding please chime in :pray:

**Comment by AlanCoding** [comment]
I have an easier time putting my remaining feelings in code here:

https://github.com/melissalkelly/awx/pull/1

While there are some simple programming opinionated bits here, the higher-level change that I am seeking is falsy handling for fields which _are valid for the type_.

As the most obtuse example, omitting the unified_job_template would be kind of crazy. There is no supported way that a job can exist without a related unified_job_template. So the idea of omitting the UJT fields when they are null doesn't make sense to me. We would rarely encounter _this_, but what will happen is that some bug occurs, and debugging is complicated by adding _another_ but on top of the first bug.

Phrasing another way, the currently implementation seems to omit the playbook field for jobs in cases, and I can not find any way to believe that might be desired.

However, this is kind of a higher-level claim related to what it is you're trying to do.

**Comment by melissalkelly** [comment]
> I have an easier time putting my remaining feelings in code here:
> 
> [melissalkelly#1](https://github.com/melissalkelly/awx/pull/1)
> 
> While there are some simple programming opinionated bits here, the higher-level change that I am seeking is falsy handling for fields which _are valid for the type_.
> 
Thanks for the feedback! You make a good point about not wanting to mask bugs. However, I do have some concerns about using `nulls` and empty strings in the payload. Since this is feedback involving architectural direction, I wanted to continue the conversation in [anstrat-1019 slack channel](https://redhat.enterprise.slack.com/archives/C0A048CSKHP). I will start a thread so we can further discuss and get other team member's thoughts. @AlanCoding

**Comment by AlanCoding** [comment]
Oh sorry, concur and created this to update

https://github.com/melissalkelly/awx/pull/2

I might need to rebase it for it to be coherent, just creating and then will try rebase

**Review by dleehr — APPROVED** [review]
Code looks good and great matching test coverage.

One thing I check for on these PRs is that the feature flag is being honored, and I don't see it being used here. I do see from the description (and from the A/C on the ticket) that we're just implementing a function that will be integrated by a future ticket. So I think that makes sense - we don't need to check the flag here, we should check the flag before calling `populate_claims_for_workload()`, correct?

**Review by AlanCoding — APPROVED** [review]
In addition to this, the way we're going to stitch things together --> I would like to add a note with an expectation of who will call `populate_claims_for_workload`, and add comments in other places this OIDC feature is expected to add to get code closure for the full feature
